### ISSUE TITLE
Preview mode broken in Next v10.0.2+ for target: "serverless": TypeError: e[t] is not a function

### ISSUE THREAD

**Issue by FinnWoelm** [vercel/next.js#20575]
# Bug report

## Describe the bug

Given any API route with `setPreviewData()`, such as the following,
```js
// pages/api/preview.js
export default async function preview(req, res) {
  res.setPreviewData({});
  res.json({ success: true });
}
```

that is built with `target: "serverless"` using Next v10.0.2 or later, will cause the following Node error when executing the page's function `default()`:

```js
TypeError: e[t] is not a function
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.293 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:29747)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.118 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:28826)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.692 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:17036)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.28 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:15901)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.27 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:3078)
TypeError: e[t] is not a function
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.293 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:29747)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.118 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:28826)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.692 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:17036)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.28 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:15901)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.27 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:3078)
(node:32970) UnhandledPromiseRejectionWarning: TypeError: e[t] is not a function
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.293 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:29747)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.118 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:28826)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.692 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:17036)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.28 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:15901)
    at __webpack_require__ (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:30130)
    at Object.27 (/home/user/next-serverless-preview-mode/.next/serverless/pages/api/preview.js:9782:3078)
```

The error occurs in the compiled API route, during `setPreviewData`, when the page attempts to `__webpack_require__` `jsonwebtoken`:
![image](https://user-images.githubusercontent.com/7606194/103293686-4e248f00-4a2b-11eb-8925-4818d1198d5b.png)

Specifically, the error occurs when `jsonwebtoken` tries to require `Buffer` (`NkYg`):
![image](https://user-images.githubusercontent.com/7606194/103293829-96dc4800-4a2b-11eb-93ec-f2ceba107fd0.png)

This is because the code tries to load `Buffer` with `__webpack_require__`, which is a function that is redefined/overwritten by the `jsonwebtoken` code:
![image](https://user-images.githubusercontent.com/7606194/103294007-f2a6d100-4a2b-11eb-8db8-c750992e01e9.png)

The root of this issue goes back to [vercel/ncc](https://github.com/vercel/ncc). The [compiled code for `jsonwebtoken`](https://github.com/vercel/next.js/blob/f06c58911515d980e25c33874c5f18ade5ac99df/packages/next/compiled/jsonwebtoken/index.js) redefines `__webpack_require__`, which takes priority over the `__webpack_require__` defined at the page-level where Buffer/`NkYg` is defined. A similar issue has been reported here: https://github.com/vercel/ncc/issues/589

This issue seems to have been introduced in `vercel/ncc` v0.24, which was introduced to `next` in v10.0.2 (https://github.com/vercel/next.js/pull/18873). In Next v10.0.1 and before, the jsonwebtoken code does **not** define `__webpack_require__` in a way that overwrites the function at the page-level.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Clone https://github.com/FinnWoelm/next-serverless-preview-mode
1. Install dependencies with `npm install`
1. Build the app with `npm run build` (or `next build`)
1. Run `node request.js`, which simulates a HTTP request and calls the API route's `default()` function
1. See error in console
1. Downgrade to Next v10.0.1: `npm install next@10.0.1`
1. Rebuild the app with `npm run build` and re-run `node request.js`
1. See no error in console and see message `{success:"true"}` in console.

## Expected behavior

Running `setPreviewData()` in an API route should not cause a fatal error. Preview mode is currently **not** usable in Next v10.0.2 with `target: "serverless"`.

## Screenshots

See screenshots included in description above.

## System information

- OS: Ubuntu 20.04
- Browser (if applies): Not applicable
- Version of Next.js: v10.0.2 and later
- Version of Node.js: tested with v12 and v14
- Deployment: any platform

## Additional context

It is much easier to trace the code and error when looking at a prettified version of the compiled API route and when looking at the prettified versions of the compiled jsonwebtoken code. I've included these in the `prettified` folder: https://github.com/FinnWoelm/next-serverless-preview-mode/tree/main/prettified

- [Prettified, compiled API route](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/main/prettified/preview.js)
  - `setPreviewData`: [L5876](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/f8afca09ddcac158d51e65eb53048b43dda98da7/prettified/preview.js#L5876)
  - requiring `jsonwebtoken`: [L5899](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/f8afca09ddcac158d51e65eb53048b43dda98da7/prettified/preview.js#L5899)
  - requiring `Buffer`: [L15472](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/f8afca09ddcac158d51e65eb53048b43dda98da7/prettified/preview.js#L15472)
  - overwriting of `__webpack_require__`: [L15492](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/f8afca09ddcac158d51e65eb53048b43dda98da7/prettified/preview.js#L15492)
- [Prettified, compiled jsonwebtoken as of Next v10.0.1](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/main/prettified/jsonwebtoken-10.0.1.js)
- [Prettified, compiled jsonwebtoken as of Next v10.0.2](https://github.com/FinnWoelm/next-serverless-preview-mode/blob/main/prettified/jsonwebtoken-10.0.2.js)

Possible workarounds:
1. Using `target: "experimental-serverless-trace"`
2. Downgrading to Next v10.0.1
3. Using `patch-package` to replace the two instances of `next/dist/compiled/jsonwebtoken` with `jsonwebtoken` in https://github.com/vercel/next.js/blob/master/packages/next/next-server/server/api-utils.ts ([patch here](https://gist.github.com/FinnWoelm/6109fdc123f293745c3424e42b260989)) and installing `jsonwebtoken` as a dependency with `npm install jsonwebtoken@3.5.1`.

Possible fix: Do not use `vercel/ncc` to precompile `jsonwebtoken` until this issue (https://github.com/vercel/ncc/issues/589) is fixed in `vercel/ncc`. This would follow the same approach that was taken with `node-fetch`, `etag`, `chalk`, and `raw-body`, which were all "un-ncc-ed" when `vercel/ncc` was upgraded to v0.25 in https://github.com/vercel/next.js/pull/18873.

PS: Huge fan of your work :hearts: Let me know if I can provide any more information.

**Comment by FinnWoelm** [comment]
I just created a PR that "un-ncc"s `jsonwebtoken`, in case it's helpful: https://github.com/vercel/next.js/pull/20577 😊 

Tested it locally (with `next: file:../next.js/packages/next`) and it works great 😊

**Comment by FinnWoelm** [comment]
It's possible that this issue would be resolved if webpack was upgraded to `v5` (to match the major version used by `vercel/ncc`). It seems to work for the issue reported here: https://github.com/vercel/ncc/issues/589. But I'm not sure if upgrading webpack to `v5` will create more problems than it will solve.

**Comment by guybedford** [comment]
Thanks for the very clear replication info here @FinnWoelm. Finally managed to track this one down with a fix in https://github.com/vercel/ncc/pull/633. As soon the ncc update is released we can reflect that build here which in my testing resolves the issue in your replication.

**Comment by rdsedmundo** [comment]
Also facing this, very frustrating because the error is not intuitive at all in pointing out what's wrong.

**Comment by FinnWoelm** [comment]
Hi @rdsedmundo, I've heard (through the grapevines from @lindsaylevine) that this is fixed in Next v10.0.6, which was released four days ago. Are you perhaps still on v10.0.5 or earlier?

**Comment by ijjk** [comment]
Per above comment it looks like this was fixed after ncc was updated so going to close this, if you're still encountering problems with this please reply with additional information and we can investigate further!
